// sns-sqs-big-payload/dist/sqs-producer.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const aws = __importStar(require("aws-sdk"));
const uuid_1 = require("uuid");
const util_1 = require("./util");
// 256KiB
exports.DEFAULT_MAX_SQS_MESSAGE_SIZE = 256 * 1024;
class SqsProducer {
    constructor(options) {
        var _a;
        if (options.sqs) {
            this.sqs = options.sqs;
        }
        else {
            this.sqs = new aws.SQS({
                region: options.region,
                endpoint: options.sqsEndpointUrl,
            });
        }
        if (options.largePayloadThoughS3 || options.allPayloadThoughS3) {
            if (!options.s3Bucket) {
                throw new Error('Need to specify "s3Bucket" option when using allPayloadThoughS3 or  largePayloadThoughS3.');
            }
            if (options.s3) {
                this.s3 = options.s3;
            }
            else {
                this.s3 = new aws.S3({
                    region: options.region,
                    endpoint: options.s3EndpointUrl,
                });
            }
        }
        this.queueUrl = options.queueUrl;
        this.largePayloadThoughS3 = options.largePayloadThoughS3;
        this.allPayloadThoughS3 = options.allPayloadThoughS3;
        this.s3Bucket = options.s3Bucket;
        this.messageSizeThreshold = (_a = options.messageSizeThreshold) !== null && _a !== void 0 ? _a : exports.DEFAULT_MAX_SQS_MESSAGE_SIZE;
        this.extendedLibraryCompatibility = options.extendedLibraryCompatibility;
    }
    static create(options) {
        return new SqsProducer(options);
    }
    async sendJSON(message, options = {}) {
        const messageBody = JSON.stringify(message);
        const msgSize = Buffer.byteLength(messageBody, 'utf-8');
        if ((msgSize > this.messageSizeThreshold && this.largePayloadThoughS3) || this.allPayloadThoughS3) {
            const payloadId = uuid_1.v4();
            const payloadKey = `${payloadId}.json`;
            const s3Response = await this.s3
                .upload({
                Bucket: this.s3Bucket,
                Body: messageBody,
                Key: payloadKey,
                ContentType: 'application/json',
            })
                .promise();
            const sqsResponse = await this.sendS3Payload({
                Id: payloadId,
                Bucket: s3Response.Bucket,
                Key: s3Response.Key,
                Location: s3Response.Location,
            }, msgSize, options);
            return {
                s3Response,
                sqsResponse,
            };
        }
        else if (msgSize > this.messageSizeThreshold) {
            throw new Error("Message is too big. Use 'largePayloadThoughS3' option to send large payloads though S3.");
        }
        const sqsResponse = await this.sqs
            .sendMessage({
            QueueUrl: this.queueUrl,
            MessageBody: messageBody,
            DelaySeconds: options.DelaySeconds,
            MessageDeduplicationId: options.MessageDeduplicationId,
            MessageGroupId: options.MessageGroupId,
        })
            .promise();
        return {
            sqsResponse,
        };
    }
    // send a message into the queue with payload which is already in s3.
    // for example: can be used to resend an unmodified message received via this lib from a queue
    // into another queue without duplicating the s3 object
    async sendS3Payload(s3PayloadMeta, msgSize, options = {}) {
        const messageAttributes = this.extendedLibraryCompatibility
            ? util_1.createExtendedCompatibilityAttributeMap(msgSize)
            : {};
        return await this.sqs
            .sendMessage({
            QueueUrl: this.queueUrl,
            MessageBody: this.extendedLibraryCompatibility
                ? util_1.buildS3PayloadWithExtendedCompatibility(s3PayloadMeta)
                : util_1.buildS3Payload(s3PayloadMeta),
            DelaySeconds: options.DelaySeconds,
            MessageDeduplicationId: options.MessageDeduplicationId,
            MessageGroupId: options.MessageGroupId,
            MessageAttributes: messageAttributes,
        })
            .promise();
    }
}
exports.SqsProducer = SqsProducer;
